src/java_bytecode/java_enum_static_init_unwind_handler.cpp

```diff
@@ -0,0 +1,61 @@
+/*******************************************************************\
+
+Module: Unwind loops in static initializers
+
+Author: Chris Smowton, chris.smowton@diffblue.com
+
+\*******************************************************************/
+
+/// \file
+/// Unwind loops in static initializers
+
+#include "java_enum_static_init_unwind_handler.h"
+
+#include <util/suffix.h>
+
+/// Unwind handler that special-cases the clinit (static initializer) functions
+/// of enumeration classes. When java_bytecode_convert_classt has annotated them
+/// with a size of the enumeration type, this forces unwinding of any loop in
+/// the static initializer to at least that many iterations, with intent to
+/// permit population of the enumeration's value array.
+/// \param loop_id: loop id, of the form function_id.loop_number
+/// \param unwind_count: iteration count that is about to begin
+/// \param [out] unwind_max: may be set to an advisory (unenforced) maximum when
+///   we know the total iteration count
+/// \param symbol_table: global symbol table
+/// \return false if loop_id belongs to an enumeration's static initializer and
+///   unwind_count is <= the enumeration size, or unknown (defer / no decision)
+///   otherwise.
+tvt java_enum_static_init_unwind_handler(
+  const irep_idt &loop_id,
+  unsigned unwind_count,
+  unsigned &unwind_max,
+  const symbol_tablet &symbol_table)
+{
+  const std::string &loop_id_string = id2string(loop_id);
+  std::size_t loop_id_delimiter_index = loop_id_string.rfind('.');
+  INVARIANT(
```

Comment:
While we're refactoring this, perhaps we could enforce this invariant with the type system (rather than passing around an `irep_idt loop_id`, we pass around a `loop_idt` which has a `function` and a `loop_index`?